### ISSUE TITLE
Some notifications are duplicated

### ISSUE THREAD

**Issue by socialError** [OneSignal/OneSignal-Android-SDK#1419]
**Description:**

In our news app, lately we have weird problem that some notifications are duplicated. It doesn't happen a lot, but I managed  to get logs with more info and it looks like there is some problem in android sdk. I tried to find out what causes this issue, but i was kind of lost. Based on provided log problem is that notification with notificationID=742dfbd9-6952-470c-a84e-006b90c0ced7 comes to OSRemoteNotificationReceivedHandler remoteNotificationReceived method twice and is also shown twice (when device is restarted etc, also both of the same notifications are restored ). In logs is also my log line with `OSRemoteNotificationReceivedHandler remoteNotificationReceived called, notificationId: 742dfbd9-6952-470c-a84e-006b90c0ced7` from remoteNotificationReceived method so you can see when was that method called.

Here is log:
https://www.dropbox.com/s/ax7tw4ro8kz7z6c/onesignal_duplicated_notification.log?dl=0

Here is my remoteNotificationReceived method. I believe I don't do anything wrong here, which could causes this issue.
```
@Override
    public void remoteNotificationReceived(Context context, OSNotificationReceivedEvent notificationReceivedEvent) {
        OSNotification notification = notificationReceivedEvent.getNotification();
        
        JSONObject data = notification.getAdditionalData();
        OSMutableNotification mutableNotification = notification.mutableCopy();
        
        if(data != null && data.has(OneSignalNotificationOpener.PAYLOAD_TAG)) {
            try {
                String topic = data.getString(OneSignalNotificationOpener.PAYLOAD_TAG);
                mutableNotification.setExtender(builder -> builder.setSubText(topic));
            }
            catch (JSONException e) {
                FirebaseCrashlytics.getInstance().recordException(e);
                e.printStackTrace();
            }
        }

        notificationReceivedEvent.complete(mutableNotification);

        Log.e("onesignal", " OSRemoteNotificationReceivedHandler remoteNotificationReceived called, notificationId: " +  mutableNotification.getNotificationId() + " androidNotificationId: " +  mutableNotification.getAndroidNotificationId() + " body: " + mutableNotification.getBody());
    }
```

If you need any more info from me, feel free to ask. We really need to solve this issue somehow. Thanks a lot

**Environment**

1. 4.4.2
2. OneSignal SDK added using the official docs (mavenCentral)

**Steps to Reproduce Issue:**

I have no steps to reproduce it, it happens completely random for us.

**Comment by socialError** [comment]
> **1 - Did this start happening about a month ago when you first reported it here: https://github.com/OneSignal/OneSignal-
> Android-SDK/issues/1419
> 2 - Can you share what changes/updates were made to your app around the time you started seeing this?**
> 

First time we saw this issue was in march 2021 when we made some changes to notifications and upgraded project libraries. We upgraded from version `3.10.9` to `4.2.0` (this issue still present in latest 4.6.0). Today to test this on sdk 3.x I modified our app and changed onesignal version to 3.10.9 and no duplicities happened yet. So it looks like this issue was created in onesignal  sdk 4.0.0. Tommorow I will test 3.15.3 and 4.0.0 versions to make sure.

**Important observation from testing for you**: It looks like duplicities happens when device is 'sleeping' and notification is shown later when device is unlocked and waken up. Not every notification is duplicated, but this duplicity happened to me always when I killed the app and left my phone on table untouched and notification came later when phone was waken up. Also I didn't turn off battery optimalization for app on my devices.




> **3 - Can you also share links to the notifications that are duplicated on your end?**

It's rare when this duplication happens, today 20 notifications were ok, and only one was duplicated.

Notification which was duplicated today during testing:
https://app.onesignal.com/apps/153b15a7-c86f-4684-a45f-fc446d1e3d54/notifications/a29dcde7-91c9-4abb-8e3f-f9d0d4ddbfab#outcomes=os__[…]ration__count
My device player ID:
3d0f225e-fd03-42df-9209-10425380a48d
My device Tags:
{user: 150457, cat:430: 10, cat:431: 10, cat:432: 10, cat:433: 10, cat:435: 10, cat:508: 10, cat:545: 10, cat:2386: 10, cat:6367: 10, cat:7657: 10, tag:3656: 30, tag:4969: 30, tag:5898: 30, tag:6439: 30, tag:6997: 30, tag:7179: 30, tag:7648: 30, tag:8562: 30, tag:8918: 30, tag:9422: 30, tag:9423: 30, tag:9424: 30, tag:9999: 30, cat:12197: 10, tag:10002: 30, tag:10038: 30, tag:10215: 30, tag:10363: 30, tag:12046: 30, tag:12047: 30, tag:12428: 30, tag:13812: 30, tag:14024: 30} (edited) 



![device-2021-09-27-134310](https://user-images.githubusercontent.com/32396097/134975156-27d5b97d-a570-4f5f-af75-a0e4f9a30003.png)

> 
**4 - Do these devices have the Minuta SK and Minuta CZ apps? Do you see the duplicates on both apps or just one of them?**

This issue happens on both apps and it doesn't matter if both or only one of them is installed on device. Also it's not device specific issue, because it's happening on different devices (tested on stock android 10, stock android 11, samsung android 8, samsung android 11).

**Comment by socialError** [comment]
Today I was testing sdk 4.0.0 version and same issue is present. So it looks like this issue was introduced in 4.0.0 version

Also I found steps how it happens a lot on production:
1. turn of wifi for some time
2. during that time server sends xy notifications
2. turn on wifi
3. notifications which were send from server during the time wifi was off will come correctly without duplicates
4. open app and kill app
5. then some duplicates of original notifications will be shown

Step 4 is maybe not mandatory, but with this steps I can reproduce it a lot. Maybe 50 percent chance to get some duplicates.

I hope it helps to locate and fix this issue.

**Comment by nan-li** [comment]
@socialError Thank you for sharing your observations, testing, and details. We will start with trying to reproduce this on our end using your steps and look into what may be happening from 3.x.x. to 4.x.x. that is affecting this.

**Comment by akushnarev-soul** [comment]
The same thing - receiving duplicates of notifications

**Comment by nan-li** [comment]
Hi @akushnarev-soul Can you share any more information about your duplicated notifications? For example:
- Are you able to reproduce it yourself?
- Are the notifications sent using filters or segments?
- Are you using `OSRemoteNotificationReceivedHandler remoteNotificationReceived`?
- When did you start seeing duplication? Any information about previous SDK version you used and which you are using now?
- Do you have any logging information to share?

Any information you have to share will be very welcome and helpful.

**Comment by akushnarev-soul** [comment]
@nan-li hi!
> Are you able to reproduce it yourself?
No, I can't, it's happening randomly.
> Are you using OSRemoteNotificationReceivedHandler remoteNotificationReceived?
yes, I see several invocations of this function.
> When did you start seeing duplication? Any information about previous SDK version you used and which you are using now?
Now we are using v4.6.0, possibly issue appeared on this version because we noticed the presence of duplicates several weeks ago

**Comment by nan-li** [comment]
@akushnarev-soul Do you recall what version you used before 4.6.0? 
I'm trying to figure out the last version you used that didn't have duplicates for you.

**Comment by akushnarev-soul** [comment]
@nan-li we used 4.5.1 before. I'm not sure that problem appeared in latest version but we noticed this issue on 4.6.0

**Comment by nan-li** [comment]
Hi @socialError,
I am looking again at your log you shared with Dropbox and I am looking for the line **"Receive receipt ending with success callback completer"** printed in the log, and it is not there. This line should print when the `NotificationWorker` is done successfully. 

Do you see this line in any other logs you have where the notifications are duplicated?

**Comment by socialError** [comment]
Log when notification is shown without duplicate:
https://www.dropbox.com/s/7v3p60g5lxfuq51/notif_correct.rtf?dl=0

Log when notification is duplicated:
https://www.dropbox.com/s/r0xaxwp81hfglsn/notif_duplicated.rtf?dl=0

Based on testing `Receive receipt ending with success callback completer:` is logged few seconds after notif is shown. For instance:
```
12:19:08.250 - OSNotificationWorkManager enqueueing notification work with notificationId: 92cefd88-df0e-4659-a8d2-d6a4ad31565c 
12:19:08.340 - OneSignal SessionManager onNotificationReceived notificationId: 92cefd88-df0e-4659-a8d2-d6a4ad31565c
12:19:21.638 - Receive receipt sent for notificationID: 92cefd88-df0e-4659-a8d2-d6a4ad31565c
12:19:21.638 - Receive receipt ending with success callback completer: androidx.concurrent.futures.CallbackToFutureAdapter$Completer@9956b3c
```
It looks like if you open app and kill app during the time between `onNotificationReceived` log and `Receive receipt ending with success callback completer:` log, notification will be processed again and will be duplicated. 

Steps how to reproduce after your input:
1. kill application
2. turn of wifi for some time
3. during that time server sends notification
4. turn on wifi - notification will be shown 
5. while `Receive receipt ending with success callback completer:` is **not** called yet open and kill app
6. then same notif will be processed again and duplicated

**Comment by nan-li** [comment]
> Today I was testing sdk 4.0.0 version and same issue is present. So it looks like this issue was introduced in 4.0.0 version

Hi @socialError, we are working on a fix but for a change made in `4.4.1`.
I am unable to reproduce on `4.0.0`. Can you share your logs for reproducing on `4.0.0`? It will be much appreciated.

**Comment by socialError** [comment]
@nan-li hello, when we should expect this fixed please?

**Comment by nan-li** [comment]
Hello @socialError, thank you for your patience.
I am optimistic that it can be shipped next week.

**Comment by nan-li** [comment]
Thank you for waiting, we have the fix now in [release 4.6.3](https://github.com/OneSignal/OneSignal-Android-SDK/releases/tag/4.6.3).

I am going to close this issue for now, but please comment/reopen if there are any more problems.
